core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java

```diff
@@ -226,36 +226,38 @@ public boolean canHandleObject(final Object obj, final boolean staticInit) {
             AppModel appModel,
             BuildTimeConfigurationReader.ReadResult readResult, LaunchMode launchMode, DevModeType devModeType) {
         Map<Class<? extends BooleanSupplier>, BooleanSupplier> condCache = new HashMap<>();
-
-        BooleanSupplier alwaysAddStep = () -> true;
-        for (CapabilityContract contract : appModel.getCapabilityContracts().values()) {
-            final String provider = contract.getExtension();
-            for (String capability : contract.getProvidesCapabilities()) {
-                int conditionIndex = capability.indexOf('?');
-                final String name = conditionIndex < 0 ? capability : capability.substring(0, conditionIndex);
-                BooleanSupplier buildIf = alwaysAddStep;
-                int testClassStart;
-                while (conditionIndex > 0) {
-                    final boolean inv = conditionIndex < capability.length() - 1
-                            && capability.charAt(conditionIndex + 1) == '!';
-                    testClassStart = conditionIndex + (inv ? 2 : 1);
-                    conditionIndex = capability.indexOf('?', testClassStart + 1);
-                    final String testClassName = capability
-                            .substring(testClassStart, conditionIndex > 0 ? conditionIndex : capability.length());
-                    Class<? extends BooleanSupplier> testClass;
-                    try {
-                        testClass = classLoader.loadClass(testClassName).asSubclass(BooleanSupplier.class);
-                    } catch (ClassNotFoundException e) {
-                        throw new RuntimeException("Failed to load capability condition class " + testClassName, e);
+        final BooleanSupplier alwaysAddStep = () -> true;
+        result = result.andThen(bcb -> bcb.addBuildStep(bc -> {
```

Comment:
Now that this is done as a build step, could it possibly be done as a build step *class* in `io/quarkus/deployment/steps`?

Reply:
Right, finally I got it. I added `ProvideCapabilitiesBuildStep`. I didn't want to re-implement `BooleanSupplier` instantiation code, so I extracted it to a public static method in the `ExtensionLoader`, which might not be the best idea. Which also means the suppliers will be created on every demand.  Would it make sense to have some global cache for them?

Reply:
I could probably move the step I added to the `CapabilityAggregationStep` instead.

Reply:
A global cache is reasonable.  The `ClassValue` class exists to associate values with classes in a less-leaky kind of way; that seems like the safest approach.

Reply:
@dmlloyd could you please give it another look? I introduced `BooleanSupplierFactoryBuildItem` as a global factory and cache of `BooleanSupplier`s.